Drain Account app

katspaugh opened the pull request:
This PR adds a new Safe app to transfer all assets to another address.

I've also replaced Lerna with Yarn Workspaces because it was easier to get the tests run this way.

<img width="857" alt="Screenshot 2021-04-15 at 15 55 18" src="https://user-images.githubusercontent.com/381895/114881122-002d6e80-9e03-11eb-831b-59cbda8b529d.png">

Closes #114

katspaugh commented:
I have read the CLA Document and I hereby sign the CLA

mikheevm commented on apps/drain-safe/.gitignore:
```diff
@@ -0,0 +1,24 @@
+# See https://help.github.com/articles/ignoring-files/ for more about ignoring files.
```
Do you need to ignore something or top-level .gitignore is enough?

mikheevm commented on apps/drain-safe/src/App.tsx:
```diff
@@ -0,0 +1,163 @@
+import { Button, Loader, Title, TextField, Table, EthHashInfo } from '@gnosis.pm/safe-react-components';
+import { useSafeAppsSDK } from '@gnosis.pm/safe-apps-react-sdk';
+import Web3 from 'web3';
+import {AbiItem} from 'web3-utils';
```
Seems the formatting is off a bit

mikheevm replied:
Ah the doesn't have a prettier pre-commit hook :(

mikheevm commented on apps/drain-safe/src/App.tsx:
```diff
@@ -0,0 +1,163 @@
+import Icon from './Icon';
+import Flex from './Flex';
+
+interface IAsset {
```
nitpick: we don't really use I* type of naming across the projects, would be good to stay consistent

mikheevm commented on apps/drain-safe/tsconfig.json:
```diff
@@ -0,0 +1,29 @@
+{
+  "compilerOptions": {
+    "target": "es5",
```
I know this comes from a template that needs to be updated, but if you have time: can we bump this to es2015? most of the browsers support it nowadays

mikheevm commented on apps/drain-safe/src/Container.tsx:
```diff
@@ -0,0 +1,13 @@
+import styled from 'styled-components';
+
+export default styled.form`
```
Do you know if this would be a violation of this rule? https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-anonymous-default-export.md

It can lead to some react-specific downsides: https://twitter.com/dan_abramov/status/1255229440860262400

I think in this specific case it would be shown as styled.form in the devtools, but not sure about the others 🤔

mikheevm commented on apps/drain-safe/src/Flex.tsx:
```diff
@@ -0,0 +1,6 @@
+import styled from 'styled-components';
+
+export default styled.div`
```
same question as above

mikheevm commented on apps/drain-safe/src/hooks/useServices/index.ts:
```diff
@@ -0,0 +1,40 @@
+
+export default function useServices(network: Networks): Services {
+  const [web3, setWeb3] = useState<Web3 | undefined>();
+  const [interfaceRepo, setInterfaceRepo] = useState<InterfaceRepository | undefined>();
```
`interfaceRepo` doesn't seem to be used anywhere

katspaugh replied:
It seems to be used in that file. Tbh, I just copied the hooks from TX Builder.

mikheevm commented on apps/drain-safe/src/App.tsx:
```diff
@@ -0,0 +1,163 @@
+    });
+
+    const params = {
+      safeTxGas: 500000,
```
This should be hardcoded as this may vary across users: some may have 1 token, others may have 100 tokens. Gas usage for these txs will be different

mikheevm commented on apps/drain-safe/src/App.tsx:
```diff
@@ -0,0 +1,163 @@
+    setSubmitting(false);
+  }, [sdk, assets, toAddress, web3, encodeTxData]);
+
+  const onToAddressChange = (e: React.ChangeEvent): void => {
+    const val = (e.target as HTMLInputElement).value;
```
ChangeEvent accepts a type argument for the element type
```suggestion
  const onToAddressChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    const val = e.target.value;
```

mikheevm reviewed (commented):
Good job!

katspaugh commented:
Thanks Mikhail! I've fixed the issues and ran prettier on the app folder.

katspaugh commented:
What would be the best way to make sure the transaction succeeded and the assets are transferred?
I've tried polling the /balances endpoint but it seems to take very long to update.

mikheevm commented:
> What would be the best way to make sure the transaction succeeded and the assets are transferred?
> I've tried polling the /balances endpoint but it seems to take very long to update.

I think for now showing a message like "The transaction has been created, refresh the app when it is executed" is the most weighted solution between effort/quality 😃 Polling is also fine.

Multisig wallets UX is an ongoing research topic we have. For a regular wallet, you could just use promises to wait until the transaction is executed. For a multi-sig wallet, it's a little bit trickier because the transaction is created, but the execution can take some time (as other people could have to sign it)

katspaugh commented:
Thanks for the detailed answer, @mikheevm! Sounds reasonable.
I'll add the message you suggested.

mikheevm commented on apps/drain-safe/src/App.tsx:
```diff
@@ -0,0 +1,209 @@
+        <EthHashInfo hash={safe.safeAddress} network={safe.network} textSize="lg" showIdenticon showCopyBtn />
+      </Flex>
+
+      {finishedTxHash && (
```
The transaction hash returned is a safe transaction hash, not the actual ethereum transaction hash, thus etherscan won't work. More info: https://docs.gnosis.io/safe/docs/contracts_tx_execution/#transaction-hash

Think of it as of internal transaction hash used by safe contracts to distinguish transactions. Also, when you approve a transaction in the interface, you actually approve the hash.

katspaugh replied:
Oh, OK. I guess it doesn't make sense to display it then?

mikheevm replied:
Yes, etherscan link will be broken and confusing for the user. Only nerds will get it 😀

katspaugh commented:
@lukasschor any critique/suggestions from your side?

lukasschor commented:
Works very well and is definitely a valuable app! Few comments

1) The icon you used is open source / free to use, right?

2) The Safe address in the top-right is a bit redundant as it's already displayed in the sidebar. Is there a reason that it is again displayed there?

3) I would personally prefer a contained button at the bottom as it is the primary action and currently not really prominent. The style you used is usually applied to secondary actions in the app.

4) When I trigger a transaction and cancel the modal, the loading spinner will disappear and the input field is available again. Which is what I would expect. However, when I trigger a transaction and select one of the sub-transactions in the modal and only then close the modal, the loading spinner is spinning indefinitely. Not sure if this is an issue related to the app or that we do not communicate the closed modal via the SDK in case the user is in the detail view of one transaction (could be as we only launched this feature recently). Not that big of an issue though, as there is the cancel button.

<img width="712" alt="CleanShot 2021-04-16 at 15 49 28@2x" src="https://user-images.githubusercontent.com/17196842/115034023-5665e480-9ecb-11eb-8feb-0646787a251a.png">

Overall, nothing that I think necessarily needs to be addressed before merging this. Just some improvement suggestions.

katspaugh commented:
Thanks Lukas!

The icon is from here: https://www.flaticon.com/free-icon/withdraw_1682382 – I'ved added an attribution in the SVG file. ✅ 

Good point about the button, fixed! ✅

Regarding the infinite loader after closing the popup from the transaction details, I'm also not sure why this isn't rejecting the send promise. Might need to investigate in the SDK.

mikheevm commented:
> Regarding the infinite loader after closing the popup from the transaction details, I'm also not sure why this isn't rejecting the send promise. Might need to investigate in the SDK.

This sounds like a bug in the integration on safe-react side. Could you please create a ticket so we don't forget about this issue?

katspaugh commented:
Created https://github.com/gnosis/safe-react/issues/2178 👍

katspaugh commented:
Feedback from @rmeissner: need to add some tests for the logic to make the app production-ready.

nicosampler commented on apps/drain-safe/src/App.tsx:
```diff
@@ -0,0 +1,192 @@
+              size="lg"
+              color="primary"
+              variant="contained"
+              onClick={submitTx}
```
does it make sense to add a confirmation modal?

mikheevm replied:
no, it looks confusing because right after that safe-react will show a confirmation modal

nicosampler replied:
that's correct. I forgot about that.

nicosampler commented on apps/drain-safe/src/App.tsx:
```diff
@@ -0,0 +1,192 @@
+  const submitTx = useCallback(async () => {
+    setSubmitting(true);
+    setFinished(false);
+
```
can we validate the address here? 
perhaps by calling `web3Utils.toChecksumAddress(toAddress)` and if it fails show an error in the input.

katspaugh replied:
Absolutely, good point 👌

nicosampler commented on apps/drain-safe/src/App.tsx:
```diff
@@ -173,19 +193,13 @@ const App: React.FC = () => {
           <TextField onChange={onToAddressChange} value={toAddress} label="Recipient" />
```
nice, remember that `TextField` already handles errors. For this case isn't neccessary because you are handling a global error. 
https://safe-components.dev.gnosisdev.com/?path=/docs/inputs-textfield--simple-text-field#error

dasanra replied:
For this you might need latest version of safe-react-components as I mentioned above

dasanra commented on apps/drain-safe/package.json:
```diff
@@ -0,0 +1,57 @@
+  "private": true,
+  "dependencies": {
+    "@gnosis.pm/safe-apps-react-sdk": "2.2.0",
+    "@gnosis.pm/safe-react-components": "^0.3.0",
```
This makes me remember that maybe we should upgrade the template to safe-react-components v0.5.0 that it's the latest one. And maybe we should release v0.6.0 shortly, there are many improvements since then

dasanra commented on apps/drain-safe/public/index.html:
```diff
@@ -0,0 +1,40 @@
+      work correctly both with client-side routing and a non-root public URL.
+      Learn how to configure a non-root public URL by running `npm run build`.
+    -->
+    <title>Gnosis Safe App</title>
```
Could we update the title here to something meaningful like `Drain Account app`?

dasanra commented on apps/drain-safe/src/App.tsx:
```diff
@@ -0,0 +1,174 @@
+    // Fetch safe assets
+    try {
+      const data = await fetchSafeAssets(safe.safeAddress, safe.network);
+      console.log('Safe balances', data);
```
Is this left on purpose?

dasanra commented on apps/drain-safe/src/App.tsx:
```diff
@@ -0,0 +1,174 @@
+    try {
+      const data = await sdk.txs.send({ txs });
+      safeTxHash = data.safeTxHash;
+      console.log(safeTxHash);
```
Is this left on purpose?

katspaugh commented:
Thanks @dasanra! Done ✅

dasanra commented on apps/drain-safe/package.json:
```diff
@@ -0,0 +1,57 @@
+    "@types/node": "^14.6.2",
+    "@types/react": "^17.0.0",
+    "@types/react-dom": "^17.0.0",
+    "@types/styled-components": "^5.1.2",
```
Would it make sense to organize this better splitting the ones that should be on devDependencies?

As now this is part of a mono-repo, should we remove those that are already in the common package.json and have no version differences?

dasanra commented on apps/drain-safe/package.json:
```diff
@@ -0,0 +1,57 @@
+  "dependencies": {
+    "@gnosis.pm/safe-apps-react-sdk": "2.2.0",
+    "@gnosis.pm/safe-react-components": "^0.5.0",
+    "@material-ui/core": "^4.11.0",
```
I think there is no direct use of this library in this app, so we could leave the version managing to safe-react-components?

mikheevm replied:
it's a peer dependency: https://github.com/gnosis/safe-react-components/blob/development/package.json#L85

mikheevm replied:
so it has to be present

katspaugh replied:
Restored 👌

dasanra reviewed (approved):
Just some small optimization comments. Overall very good work here :+1:

katspaugh commented:
Thanks @dasanra! I've removed the console logs and cleaned up the dependencies. 👍

rmeissner commented on apps/drain-safe/src/utils/gateway.ts:
```diff
@@ -0,0 +1,29 @@
+
+export async function fetchSafeAssets(safeAddress: string, safeNetwork: string): Promise<Balance> {
+  const network = safeNetwork.toLowerCase() === 'mainnet' ? 'mainnet' : 'rinkeby';
+  const url = `https://safe-client.${network}.gnosis.io/v1/safes/${safeAddress}/balances/${CURRENCY}/?trusted=false&exclude_spam=true`;
```
While this is indeed the easiest way, be aware that the client gateway is for the client, so we will deprecate endpoints that are not used by the mobile clients or react interface. We should find a better way for this in the future. I don't think it is blocking for now.

katspaugh replied:
I'll check if I can use the transactions API directly.

rmeissner replied:
What I meant is that theoretically you can use events to determine what tokens have been used with this address and fetch the balances for these tokens (that would be the version that does not rely on our services at all) :D

Again I don't think this is required for an initial version.

katspaugh commented on .github/workflows/deploy.pr.yml:
```diff
@@ -48,7 +48,7 @@ jobs:
 
       - name: Build apps
-        run: npx lerna run build
+        run: yarn build
```
Tests weren't being run properly as far as I could tell, so I "fixed" it by removing Lerna. Not a problem to restore if you think we'll need it in future.

rmeissner replied:
I don'T really mind. @mikheevm set this up so he should know better, I personally prefer less dependencies, so if lerna is not required feel free to remove it :)

katspaugh commented on .github/workflows/deployment.yml:
```diff
@@ -51,7 +51,7 @@ jobs:
 
       - name: Build apps
-        run: npx lerna run build
+        run: yarn build
```
Added ✅

katspaugh commented:
@rmeissner I've addressed your comments, thank you!

The test and lint jobs are failing with this error:
> error https://registry.yarnpkg.com/lodash/-/lodash-4.17.20.tgz: Extracting tar content of undefined failed, the file appears to be corrupt: "ENOENT: no such file or directory, stat '/usr/local/share/.cache/yarn/v6/npm-lodash-4.17.20-b44a9b6297bcb698f1c51a3545a2b3b368d59c52-integrity/node_modules/lodash/fp/compact.js'"

Looks like something on the Yarn repo side? I'll try to rebuild later today.

rmeissner commented on apps/drain-safe/src/utils/gateway.ts:
```diff
@@ -0,0 +1,26 @@
+
+export async function fetchSafeAssets(safeAddress: string, safeNetwork: string): Promise<Asset[]> {
+  const network = safeNetwork.toLowerCase(); // mainnet, rinkeby etc
+  const url = `https://safe-transaction.${network}.gnosis.io/api/v1/safes/${safeAddress}/balances/${CURRENCY.toLowerCase()}/?exclude_spam=true`;
```
You have to be aware that on some networks this might not be available, so in that case the request will fail ;)

Minor nitpick: the file is called `gateway.ts` but we use don't use the gateway here :P

katspaugh replied:
Renamed 👍

liliya-soroka commented:
@katspaugh , should it go through QA? It was not tested and approved by QA

liliya-soroka commented:
Not approved by QA:

- [ ] Decimals are not taken into account for the token value:
Drain account app :
![image.png](https://images.zenhubusercontent.com/5ed757c29459bb36bacd4bba/3b47d76f-8053-4bae-9014-227547087481)
Actual values on balance page where decimals are taken into account:
![image.png](https://images.zenhubusercontent.com/5ed757c29459bb36bacd4bba/0ab209d7-5c83-4304-ac50-94f8d941099d)
![image.png](https://images.zenhubusercontent.com/5ed757c29459bb36bacd4bba/130102f2-ff97-4ae3-b2c6-afda59769002)

- [ ] no default icon for the tokens without icon as on balance page :
Current result: ![image.png](https://images.zenhubusercontent.com/5ed757c29459bb36bacd4bba/b00c52ad-aa20-40c1-a3ab-c49599217e57)__
Expected result: ![image.png](https://images.zenhubusercontent.com/5ed757c29459bb36bacd4bba/cc7ca2de-ee47-4883-879d-5d26a991624d)

- [ ] What is the default sorting ?
![image.png](https://images.zenhubusercontent.com/5ed757c29459bb36bacd4bba/ebebe054-76ac-44df-ba0a-8d0caf772989)

katspaugh commented:
@liliya-soroka I thought we were supposed to test individual tickets ourselves and you only to do the regression.

Anyway, thanks for testing so quickly! Could you create a separate issue for everything together?
Except for the last point with the sorting arrow – it's an issue with the Table component from SRC.

bh2smith commented on apps/drain-safe/.env.sample:
```diff
@@ -0,0 +1 @@
+PORT=3000
```
nanonit: newline at EOF.

katspaugh replied:
Thanks Ben 😄  Fixed in #126.

liliya-soroka commented:
Verified

katspaugh commented:
Thanks @liliya-soroka!
